rust_src/src/terminal.rs

```diff
@@ -0,0 +1,83 @@
+//! Functions related to terminal devices.
+
+use std::{mem, ptr};
+
+use libc::c_void;
+
+use remacs_macros::lisp_fn;
+use remacs_sys::build_string;
+use remacs_sys::{Lisp_Terminal, Qterminal_live_p};
+
+use frames::selected_frame;
+use lisp::defsubr;
+use lisp::{ExternalPtr, LispObject};
+
+pub type LispTerminalRef = ExternalPtr<Lisp_Terminal>;
+
+impl LispTerminalRef {
+    pub fn from_ptr(ptr: *mut c_void) -> Option<LispTerminalRef> {
+        unsafe { ptr.as_ref().map(|p| mem::transmute(p)) }
+    }
+
+    pub fn is_live(self) -> bool {
+        !self.name.is_null()
+    }
+
+    pub fn name(self) -> LispObject {
+        unsafe { build_string(self.name) }
+    }
+}
+
+/// Return the terminal object specified by TERMINAL.  TERMINAL may
+/// be a terminal object, a frame, or nil for the terminal device of
+/// the current frame.  If TERMINAL is neither from the above or the
+/// resulting terminal object is deleted, return NULL.
+#[no_mangle]
+pub extern "C" fn decode_terminal(mut terminal: LispObject) -> *mut Lisp_Terminal {
+    if terminal.is_nil() {
+        terminal = selected_frame();
+    }
+
+    let t;
+    if let Some(mut term) = terminal.as_terminal() {
+        t = term.as_mut();
+    } else if let Some(frame) = terminal.as_frame() {
+        t = frame.terminal;
+    } else {
+        t = ptr::null_mut();
+    }
+
+    if let Some(term_ref) = LispTerminalRef::from_ptr(t as *mut c_void) {
```

Comment:
I guess this causes segfault
We should check for null before transmuting to `LispTerminalRef`

Reply:
We could assert that pointer is not null in `from_prt` methods too

Reply:
Oh, I thought this is handled by `as_ref`.

Reply:
`ptr.as_ref` handles whether the pre is null. That is why the return value is an Option type.

Reply:
https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref